### DISCUSSION TITLE
Use RabbitMQ behind an HTTP proxy

### DISCUSSION THREAD

**Discussion by AdelAljalam** [rabbitmq/rabbitmq-dotnet-client#1031] [Q&A]
Hi all,

I have a .NET service that uses the RabbitMQ .NET client to connect to a remote RabbitMQ server in the cloud. That works fine when no proxy is used in the local machine where the service resides. In some cases, a proxy is present preventing all other traffic than HTTP(s) and blocking all ports other than 80 and 443. In that case, I have a problem connecting to the RabbitMQ server. Till now, I found no direct way to make my service work behind a proxy. 

Trying to investigate the issue, I found the following:

- This [pull request](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/115/commits/7673771d3ab0ffb6633117aeab6f3d92252bc3ca) The code is supposed to tunnel AMQP traffic into an HTTP(s) proxy. The pull request seems to be integrated in some old version of the client and then removed for some unknown reason. Can someone confirm whether this code is supposed to solve my problem? Did someone use it before? Is it well tested as it was removed later on from newer versions?
- Enable RabbitMQ HTTP(s) management plugin in order to send HTTP messages instead of AMQP ones. Is it a good idea in your opinion as I will lose all AMQP protocol advantages?
- One another solution related to proxying is to enable the [proxy protocol ](https://www.rabbitmq.com/networking.html#proxy-protocol) in RabbitMQ server. It seems to be a TCP proxy and in addition, it is not supported in the RabbitMQ .NET client so the protocol should be implemented if I understand well. Can somebody confirm please?

Does someone have another solution or did already face this type of issues?

Best,
Adel

**Comment by michaelklishin** [comment]
`TunneledTcpClient` has been removed because maintaining it was much more painful than "regular" TCP sockets and it hasn't seen much used.

RabbitMQ core team does not comment using HTTP API for publishing and consumption. It's going to be very inefficient because it cannot [maintain a long lived messaging protocol connection](https://www.rabbitmq.com/connections.html).

Proxy protocol is not a TCP proxy. It's a feature that servers and proxies support that reveals real client IP address instead of the proxy address to the upstream service (such as RabbitMQ). No client support of any kind is needed as this feature
almost by definition involves using a proxy.

I don't have a lot to recommend other than [exposing a messaging protocol port and using that](https://www.rabbitmq.com/networking.html#ports). FWIW RabbitMQ supports pretty standard and well-known protocols that have been around for a decade or more.

**Reply by AdelAljalam** [reply]
Thanks for the information about the proxy protocol. It was not so clear for me as it is not that much documented.

Unfortunately, what you recommend is something we want to avoid as we have thousands of RabbitMQ clients connecting from different machines with different firewall and proxying policies, it would be painful to expose ports for all of them.
